File: aoc2021/day1_sonar_sweep/main.py
import os, sys

task_dir = os.path.dirname(__file__)
sys.path.append(f"{task_dir}/..")
from get_tasks import get_input

measures, example = get_input(task_dir, 1)
# print(example)


def part1(input):
    counter = 0
    temp = max(int(m) for m in input)
    for measure in input:
        if int(measure) > temp:
            counter += 1
        temp = int(measure)

    print("The answer of part1 is:", counter)


def check_part1(example):
    part1(example)


def part2(input):
    input = [int(m) for m in input]
    counter = 0
    temp = sum(input[0:3])
    for ind in range(len(input) - 2):
        if sum(input[ind : ind + 3]) > temp:
            counter += 1
        temp = sum(input[ind : ind + 3])

    print("The answer of part2 is:", counter)


def check_part2(example):
    part2(example)


if __name__ == "__main__":
    check_part1(example)
    part1(measures)
    check_part2(example)
    part2(measures)

File: aoc2021/day1_sonar_sweep/src/main.rs
fn main() {
    let measures: Vec<i32> = include_str!("../input.txt")
        .lines()
        .map(|x| x.parse().unwrap())
        .collect();
    let example: Vec<i32> = include_str!("../example.txt")
        .lines()
        .map(|x| x.parse().unwrap())
        .collect();

    check_example(&example, &part1);
    check_example(&example, &part2);
    part1(&measures);
    part2(&measures);
}

fn check_example(example: &Vec<i32>, part: &dyn Fn(&Vec<i32>)) {
    part(&example)
}

fn part1(input: &Vec<i32>) {
    let mut counter = 0;
    let mut temp = input.iter().max().unwrap();
    for measure in input {
        if measure > temp {
            counter += 1;
        }
        temp = measure
    }
    println!("The Answer of part1 is: {}", counter);
}

fn part2(input: &Vec<i32>) {
    let mut counter = 0;
    // let temp: i32 = input[0..3].iter().sum();
    for ind in 0..(input.len() - 3) {
        if input[ind] < input[ind + 3] {
            // println!("{}, {}, {}", ind, input[ind], input[ind + 3]);
            counter += 1
        }
    }
    println!("The Answer of part2 is: {}", counter);
}
